src/OptFrame/Heuristics/EvolutionaryAlgorithms/Election.hpp

```diff
@@ -0,0 +1,159 @@
+// OptFrame - Optimization Framework
+
+// Copyright (C) 2009-2015
+// http://optframe.sourceforge.net/
+//
+// This file is part of the OptFrame optimization framework. This framework
+// is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License v3 as published by the
+// Free Software Foundation.
+
+// This framework is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License v3 for more details.
+
+// You should have received a copy of the GNU Lesser General Public License v3
+// along with this library; see the file COPYING.  If not, write to the Free
+// Software Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
+// USA.
+
+#ifndef OPTFRAME_GENERAL_ELECTION_HPP_
+#define OPTFRAME_GENERAL_ELECTION_HPP_
+
+#include <chrono>
+#include <random>
+#include <limits>
+
+#include "../../Evaluation.hpp"
+#include "../../BaseSolution.h"
+#include "../../Util/printable.h"
+#include "../../MultiSolution.hpp"
+#include "../../MultiEvaluation.hpp"
+#include "EA.h"
+
+#ifndef _OPTFRAME_DBG_ELECTION_
+#   ifdef OPTFRAME_DEBUG
+#       define _OPTFRAME_DBG_ELECTION_ 
+#   else
+#       define _OPTFRAME_DBG_ELECTION_ while(false)
```

Comment:
What's this????

Reply:
A way to debug code. It's not as clean as a debugger but for people like us who debug code by printing its behavior its pretty clean. Usage is as follows:

DEBUG  std::cerr << "This is a single line of debug" << std::endl;
DEBUG{
     std::cerr << "Multiple";
     std::cerr << "lines";
     std::cerr << "of debug" << std::endl;
     int var; //a variable that is only visible inside this debug code
}  

The advantages and disvantages of this method are:
++Clean
++Can use local variables 
++Will be checked for errors everytime you compile, unlike pure #ifdef macros
--May generate higher executables
--May not be optimized away by some compilers (GCC, clang and intel does)

For framework users this should be the guidelines (if adopted in the entire project):

++ Don't use -DNDEBUG
++Use -DOPTFRAME_DEBUG for full framework debugging. All lines to written to err with -OptDebug- means that the framework did the printing.
++If you want to disable component-specific debugging, then you should manually define the debug line for it. For exmple, if you want to disable ELECTION debug, then you should define -DOPTFRAME_DEBUG and -D_OPTFRAME_DBG_ELECTION_.